mesh/mesh.go

```diff
@@ -500,3 +500,63 @@ func (m *Mesh) GetATXs(atxIds []types.AtxId) (map[types.AtxId]*types.ActivationT
 	}
 	return atxs, mIds
 }
+
+// ActiveSetForLayerView - returns the active set size that matches the view of the contextually valid blocks in the provided layer
+func (m *Mesh) ActiveSetForLayerView(layer types.LayerID, layersPerEpoch uint16) (uint32, error) {
+
+	epoch := layer.GetEpoch(layersPerEpoch)
+	firstLayerOfPrevEpoch := types.LayerID(epoch-1) * types.LayerID(layersPerEpoch)
+
+	// build a map of all blocks between firstLayerOfPrevEpoch and our current layer
+	mp := make(map[types.BlockID]struct{})
+	blocks, err := m.LayerBlockIds(layer)
+	for _, bid := range blocks {
+		mp[bid] = struct{}{}
+	}
+
+	activeSetSize := uint32(0)
+	countedAtxs := make(map[types.AtxId]struct{})
+
+	traversalFunc := func(blkh *types.BlockHeader) error {
+
+		// only contextually valid blocks
+		if !m.IsContexuallyValid(blkh.Id) {
+			return nil
+		}
+
+		blk, err := m.GetBlock(blkh.Id)
```

Comment:
you should have all the syntacticly valid blocks in the view
you shouldn't have to check if the block exists

Reply:
Yes, but I do however need to fetch the atxs (and the bheader isn't enough) so I also check err.